TigranKirakosov/hw4.js

```diff
@@ -0,0 +1,173 @@
+};
+Food.prototype.showPrice = function() {
+  return this.typeOfFood.tugric;
+};
+Food.prototype.showCalories = function() {
+  return this.typeOfFood.calories;
+};
+Food.prototype.showType = function() {
+  return `It is ${this.typeOfFood.type}`;
+};
+Food.prototype.showSize = function() {
+  return `${this.size.type} size`;
+};
+
+// Class Hamburger
+function Hamburger(description) {
+  Food.call(this, description);
+}
+Hamburger.prototype = Object.create(Food.prototype);
+
+// Hamburger constant vars
+Hamburger.SIZES = [
+  { type: 'Small', tugric: 50, calories: 20 },
+  { type: 'Large', tugric: 100, calories: 40 }
+];
+Hamburger.ADDITIONS = [
+  { type: 'Cheese', tugric: 10, calories: 20 },
+  { type: 'Salad', tugric: 20, calories: 5 },
+  { type: 'Potato', tugric: 15, calories: 10 }
+];
+
+// Hamburger methods
+Hamburger.prototype.showPrice = function() {
+  return this.size.tugric + this.addition.tugric;
+};
+Hamburger.prototype.showCalories = function() {
+  return this.size.calories + this.addition.calories;
+};
+
+// Class Salad
+function Salad(description) {
+  Food.call(this, description);
+}
+Salad.prototype = Object.create(Food.prototype);
+
+// Salad constant vars
+Salad.MENU = [
+  { type: 'Caesar', tugric: 100, calories: 20 },
+  { type: 'Olivier', tugric: 50, calories: 80 }
+];
+
+// Class Drink
+function Drink(description) {
+  Food.call(this, description);
+}
+Drink.prototype = Object.create(Food.prototype);
+
+// Drink constant vars
+Drink.MENU = [
+  { type: 'Cola', tugric: 50, calories: 40 },
+  { type: 'Coffee', tugric: 80, calories: 20 }
+];
+
+/* 
+#                           Order Interface
+##
+### addToOrder(arg1, arg2, ...)   || Add multiple items
+### deleteFromOrder(position)     || Delete certain position
+### showPositions()               || Show positions by its type
+### calcTotalPrice()              || Show total price of an order
+### calcTotalCalories()           || Show total calories an order provides
+### makePayment()                 || Pay an order and make it immutable
+##
+#                           Food Interface
+##
+### chooseAddition(type, menu)    || Sets addition from menu
+### chooseType(type, menu)        || Sets type from menu
+### chooseSize(type, menu)        || Sets size from menu
+### showSize()                    || Shows size type
+### showAddition()                || Shows addition type
+### showType()                    || Shows type
+### showPrice()                   || Shows price
+### showCalories()                || Shows calories
+##
+# 
+*/
+
+// Instantiating objects
+var myOrder = new Order('Tigran');
+var myHamb = new Hamburger('Hamburger');
+var mySalad = new Salad('Olivier salad');
+var myDrink = new Drink('Cola');
+
+// Configurating food properties
+myHamb.chooseSize('Large', Hamburger.SIZES); // 'Large' or 'Small'
+myHamb.chooseAddition('Potato', Hamburger.ADDITIONS); // 'Cheese', 'Salad' or 'Potato'
+mySalad.chooseType('Olivier', Salad.MENU); // 'Caesar' or 'Olivier'

```

Comment:
Ok, you extracted logic to Food - it's nice :), but current implementation has big problem - I can write smth like `mySalad.chooseType('Cola', Drink.MENU);` - and it'll work. The problem is that, in fact the second parameter - it's private part of each class and user shouldn't set it in public api of your classes. You need to extend logic of `Food` class in each children - smth like
 ` function Salad(description) {`
  `Food.call(this, description);`
  `this.chooseType = function(type) {`
    `Food.prototype.chooseType.call(this, type, Salad.MENU);`
`}`
`}`

Reply:
maybe this article will be helpful https://learn.javascript.ru/functional-inheritance

Reply:
Alright, I see there's a problem. Thanks for the article!